README
byCycle Trip Planner
++++++++++++++++++++

Introduction
============

The byCycle Trip Planner is a system for planning trips made by bicycle. Other
travel modes are possible but none have been implemented yet.


License and Warranty
====================

See the LICENSE file.


Installation
============

See the INSTALL file.


More Information
================

Information about the byCycle project can be found at http://info.bycycle.org/.


Contact
=======

You can contact us to ask questions, make comments, offer suggestions, get
help, offer help, etc at contact@bycycle.org.


CHANGELOG
0.5 (unreleased)
----------------

- Improved performance of Route service by using "joined loads" to load edges
  and nodes when creating directions list. The number of database queries was
  O(number of edges); now it's O(1).

- Include list of bikemodes for each direction in directions list.


0.5a1 (2012-12-07)
------------------

Made changes necessary to get things back up and running. Made a few other
changes to clean things up a bit.

- Using str06oct from the Pirate network, because that's the only data set
  I could find in backups. I have no idea how up-to-date/out-of-date it is.

- Upgraded all dependencies to latest versions; removed unused. Almost no
  changes were required by these upgrades; there was one tiny change to
  a method signature in our user-defined SQLAlchemy geometry type, and I think
  that was it... oh, and SQLError was renamed to SQLAlchemyError.

- Added glineenc dependency; removed local copy.

- Upgraded to Dijkstar 2 and simplified graph-related code. Creation of
  regional adjacency matrix is simpler; splitting of edges in Route service is
  much simpler.

- Updated INSTALL doc to reflect latest Ubuntu packages. Also cleaned it up
  some.

- Improved memory usage during matrix creation (process batches of edges
  instead of all at once).

- Converted integrate and matrix scripts to proper console scripts (using
  argparse and console_scripts entry point).

- Renamed Geocode.xy_ll to lat_long; added Route.linestring_lat_long.

- Removed Milwaukee, Wisconsin, region.

- Cleaned up tests; added more.

- Cleaned up some cruft (unused files and whatnot).

- Various other small improvements and cleanup; see `hg diff -r
  bycycle-is-back:tip` for more details.


0.4
---

- Migrated from MySQL to PostgreSQL
- Uses SQLAlchemy to interact with the database instead of using raw SQL
- Uses PostGIS for storing and operating on spatial data
- Uses the PCL to manipulate geometries in Python
- Moved services into classes that have a common base
- Switched to a reStructuredText style of docstrings
- Cleaned up the route service quite a bit
- Added a unittest test suite for the route service
- Created a Route class


Ancient History (<= 0.4)
------------------------

SVN
* 2007-04-24:
    * Made changes necessary to get normaddr service tests to pass
    * Made changes necessary to get all but one geocode service tests to pass
* 2007-04-25
    * Made changes necessary to get ALL geocode service tests to pass
* 2007-04-26
    * Made changes necessary to get ALL tests to pass, including for routes
    * Changed Node and Edge inheritance so that subclasses classes point at
      base classes (with a Foreign Key) instead of the other way around.
      Base classes now have no knowledge of subclasses.
* 2007-04-27
    * Made changes necessary to get route tests to pass AND be *meaningful*;
      previously, the tests were passing, but the edges were out of order
      due to the fact that doing a "SELECT ... IN ..." doesn't necessarily
      return rows in the same order as the IDs in the IN clause.
    * Made some incidental changes affecting routing also.
* 2007-05-01
    * Removed address unification scheme. It didn't accomplish very well what
      is was trying to accomplish, and there are better ways of locating an
      address when the number is not in range for a given street name (A: make
      sure address ranges are all correct, B: search for the edge with the
      address range that has the smallest delta from the input number).
* 2007-05-02
    * Went back to schema scheme (each region has it's own schema). This
      removes some complexity and fixes some issues with locating the
      correct attributes for edges and nodes.


byCycle/model/route.py
###############################################################################
# $Id: geocode.py 212 2006-09-11 04:16:40Z bycycle $
# Created 2006-09-25.
#
# Route entity.
#
# Copyright (C) 2006, 2007 Wyatt Baldwin, byCycle.org <wyatt@bycycle.org>.
# All rights reserved.
#
# For terms of use and warranty details, please see the LICENSE file included
# in the top level of this distribution. This software is provided AS IS with
# NO WARRANTY OF ANY KIND.
###############################################################################
"""Route entity."""
from cartography.proj import SpatialReference
from byCycle.model import glineenc

__all__ = ['Route']


class Route(object):
    """Represents a route between two addresses."""

    def __init__(self,
                 region,
                 start, end,
                 directions, linestring, distance):
        self.region = region
        self.start = start
        self.end = end
        self.directions = directions
        self.distance = distance

        linestring.srs = SpatialReference(epsg=region.srid)
        self.linestring = linestring

        linestring_ll = linestring.copy()
        ll_srs = SpatialReference(epsg=4326)
        linestring_ll.transform(src_proj=str(self.linestring.srs),
                                dst_proj=str(ll_srs))
        self.linestring_ll = linestring_ll

    def to_builtin(self):
        points = []
        for i in range(self.linestring_ll.numPoints()):
            points.append(self.linestring_ll.pointN(i))
        linestring = [{'x': p.x, 'y': p.y} for p in points]
        pairs = [(p.y, p.x) for p in points]
        bounds = self.linestring_ll.envelope()
        centroid = bounds.centroid()
        google_points, google_levels = glineenc.encode_pairs(pairs)
        route = {
            'start': dict(self.start),
            'end': dict(self.end),
            'linestring': linestring,
            'bounds': {
                'sw': {'x': bounds.minx, 'y': bounds.miny},
                'ne': {'x': bounds.maxx, 'y': bounds.maxy}
            },
            'center': {'x': centroid.x, 'y': centroid.y},
            'directions': self.directions,
            'distance': self.distance,
            'google_points': google_points,
            'google_levels': google_levels,
        }
        route['start']['geocode'] = route['start']['geocode'].to_builtin()
        route['end']['geocode'] = route['end']['geocode'].to_builtin()
        return route

    def __repr__(self):
        return repr(self.to_builtin())

    def __str__(self):
        directions = []
        for d in self.directions:
            dbm = d['bikemode']
            bm = [', '.join([[b, '-'][b is 'n'] for b in dbm]), ''][not dbm]
            directions.append('%s on %s toward %s -- %s %s [%s]' % (
                d['turn'],
                d['street'],
                d['toward'],
                '%.2f' % (d['distance']['miles']),
                'miles',
                bm,
            ))
        directions = '\n'.join([
            '%s%s. %s' % (['', ' '][i<10], i, d)
            for i, d
            in enumerate(directions)
        ])
        s = [
            self.start['geocode'],
            self.end['geocode'],
            'Distance: %.2f' % (self.distance['miles']),
            directions,
        ]
        return '\n'.join([str(a) for a in s])
